src/mtt_internals.c

```diff
@@ -103,8 +104,24 @@ MEMKIND_EXPORT void *mtt_internals_realloc(MttInternals *internals, void *ptr,
 {
     // TODO this is more complicated: realloc might call malloc, so we actually
     // need to extend pool
+
+    if (size == 0) {
+        mtt_internals_free(internals, ptr);
+        return NULL;
+    } else if (ptr == NULL) {
+        return mtt_internals_malloc(internals, size);
+    }
+
+    size_t old_size = mtt_internals_usable_size(internals, ptr);
+    if (size <= old_size) {
```

Comment:
What if user allocated e.g. 1GB and then reallocates to 1kB ? We would waste a lot of memory.